java/core/src/test/java/com/microsoft/bond/codegen/EqualsHashCodeTests.java

```diff
@@ -0,0 +1,185 @@
+        assertObjectsAreEqual(r, r);
+        assertObjectsAreEqual(d, d);
+    }
+
+    @Test
+    public void defaultObjects() {
+        assertObjectsAreEqual(new Empty(), new Empty());
+        assertObjectsAreEqual(new Primitives(), new Primitives());
+        assertObjectsAreEqual(new Containers(), new Containers());
+        assertObjectsAreEqual(new Recursive(), new Recursive());
+        assertObjectsAreEqual(new Derived(), new Derived());
+    }
+
+    @Test
+    public void primitiveFields() {
+        final Primitives one = new Primitives(), another = new Primitives();
+        for (final Primitives p : Arrays.asList(one, another)) {
+            p.iint8 = 3;
+            p.iint16 = 3;
+            p.iint32 = 3;
+            p.iint64 = 3;
+            p.iuint8 = 3;
+            p.iuint16 = 3;
+            p.iuint32 = 3;
+            p.iuint64 = 3;
+            p.ffloat = 3.0f;
+            p.fdouble = 3.0;
+            p.bbool = true;
+            p.str = "three";
+            p.wstr = "three";
+        }
+        assertObjectsAreEqual(one, another);
+
+        // Break each field's equality.
+        another.iint8 = (byte) (one.iint8 + 1);
+        Assert.assertNotEquals(one, another);
+        another.iint8 = one.iint8;
+
+        another.iint16 = (short) (one.iint16 + 1);
+        Assert.assertNotEquals(one, another);
+        another.iint16 = one.iint16;
+
+        another.iint32 = one.iint32 + 1;
+        Assert.assertNotEquals(one, another);
+        another.iint32 = one.iint32;
+
+        another.iint64 = one.iint64 + 1;
+        Assert.assertNotEquals(one, another);
+        another.iint64 = one.iint64;
+
+        another.iuint8 = (byte) (one.iuint8 + 1);
+        Assert.assertNotEquals(one, another);
+        another.iuint8 = one.iuint8;
+
+        another.iuint16 = (short) (one.iuint16 + 1);
+        Assert.assertNotEquals(one, another);
+        another.iuint16 = one.iuint16;
+
+        another.iuint32 = one.iuint32 + 1;
+        Assert.assertNotEquals(one, another);
+        another.iuint32 = one.iuint32;
+
+        another.iuint64 = one.iuint64 + 1;
+        Assert.assertNotEquals(one, another);
+        another.iuint64 = one.iuint64;
+
+        another.ffloat = one.ffloat + 1.0f;
+        Assert.assertNotEquals(one, another);
+        another.ffloat = one.ffloat;
+
+        another.fdouble = one.fdouble + 1.0f;
+        Assert.assertNotEquals(one, another);
+        another.fdouble = one.fdouble;
+
+        another.bbool = !one.bbool;
+        Assert.assertNotEquals(one, another);
+        another.bbool = one.bbool;
+
+        another.str = one.str + "eee";
+        Assert.assertNotEquals(one, another);
+        another.str = one.str;
+
+        another.wstr = one.wstr + "eee";
+        Assert.assertNotEquals(one, another);
+        another.wstr = one.wstr;
+    }
+
+    @Test
+    public void containerFields() {
+        final Containers one = new Containers(), another = new Containers();
```

Comment:
Consider adding tests for containers with struct values to ensure that structural equality is recursively applied to the elements of containers.